curp/src/server/curp_node.rs

```diff
@@ -490,49 +498,92 @@ impl<C: 'static + Command> CurpNode<C> {
                     }
                 }
             };
-            tokio::time::sleep(retry_timeout).await;
+            sleep(retry_timeout).await;
         }
     }
 
-    /// Sync task is responsible for replicating log entries
-    async fn sync_task(
-        curp: Arc<RawCurp<C>>,
-        connects: HashMap<ServerId, Arc<impl ConnectApi>>,
-        mut sync_rx: mpsc::UnboundedReceiver<LogIndex>,
+    /// invoke `send_log_to_follower` for every follower
+    fn bcast_append_entries(
+        curp: &Arc<RawCurp<C>>,
+        connects: &HashMap<ServerId, Arc<impl ConnectApi>>,
+        batch: Vec<LogIndex>,
     ) {
-        while let Some(i) = sync_rx.recv().await {
-            let req = {
-                let Ok(ae) = curp.append_entries_single(i) else {
-                    error!("{} get log {i} failed", curp.id());
-                    continue;
-                };
-                let req = AppendEntriesRequest::new(
-                    ae.term,
-                    ae.leader_id,
-                    ae.prev_log_index,
-                    ae.prev_log_term,
-                    ae.entries,
-                    ae.leader_commit,
-                );
-                match req {
-                    Err(e) => {
-                        error!("can't serialize log entries, {e}");
-                        continue;
-                    }
-                    Ok(req) => req,
+        if let Some(&last_sent_idx) = batch.last() {
+            let Ok(ae) = curp.append_entries_batch(batch) else {
+                return;
+            };
+
+            let req = match AppendEntriesRequest::new(
+                ae.term,
+                ae.leader_id,
+                ae.prev_log_index,
+                ae.prev_log_term,
+                ae.entries,
+                ae.leader_commit,
+            ) {
+                Err(e) => {
+                    error!("unable to serialize append entries request: {}", e);
+                    return;
                 }
+                Ok(req) => req,
             };
+
             for connect in connects.values().cloned() {
                 let _handle = tokio::spawn(Self::send_log_to_follower(
-                    Arc::clone(&curp),
+                    Arc::clone(curp),
                     connect,
-                    i,
+                    last_sent_idx,
                     req.clone(),
                 ));
             }
         }
     }
 
+    /// Sync task is responsible for replicating log entries
+    async fn sync_task(
+        curp: Arc<RawCurp<C>>,
+        connects: HashMap<ServerId, Arc<impl ConnectApi>>,
+        sync_rx: flume::Receiver<(LogIndex, usize)>,
+    ) {
+        let batch_timeout = curp.cfg().batch_timeout;
+        let batch_max_size = curp.cfg().batch_max_size;
+        let connects = Arc::new(connects);
+        let timer = sleep(batch_timeout);
+        let batch = Mutex::new(Vec::new());
+        let batch_size = AtomicUsize::new(0);
```

Comment:
Why do you use atomic number here? I don't think it'll be accessed parallelly.

Reply:
Fixed.